dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,369 @@
+  out.reserve(sample_config_num);
+
+  for (size_t sample_index = 0; sample_index < sample_config_num; sample_index++) {
+    SampleConfig new_sample = ParseSampleConfig(config, config_path);
+    if (!out.empty()) {
+      out.back().end_offset = std::min(out.back().end_offset, new_sample.start_offset);
+    }
+
+    // filtering out samples without the required extensions
+    if (std::all_of(ext_.begin(), ext_.end(), [&](std::set<std::string> extension_set) {
+          return HasIntersection(extension_set, new_sample.extensions);
+        })) {
+      switch (missing_component_behavior_) {
+        case MissingExt::Skip:
+          continue;
+        case MissingExt::Raise:
+          DALI_ERROR("Underfull sample detected in index file at " + config_path);
+        default:
+          break;
+      };
+    }
+    out.push_back(std::move(new_sample));
+  }
+  if (out.size()) {
+    out.back().end_offset = std::min(out.back().end_offset, config_size);
+  }
+  return out;
+}
+
+void WebdatasetLoader::PrepareMetadataImpl() {
+  if (!dont_use_mmap_) {
+    mmap_reserver_ = FileStream::MappingReserver(static_cast<unsigned int>(uris_.size()));
+  }
+  copy_read_data_ = dont_use_mmap_ || !mmap_reserver_.CanShareMappedData();
+
+  // reserving the data in vector fields
+  wds_shards_metadata_.reserve(configs_.size());
+  wds_shards_.reserve(uris_.size());
+  wds_shards_prefixsums_.reserve(configs_.size());
+
+  // collecting the config files
+  for (auto& config_path : configs_) {
+    wds_shards_metadata_.emplace_back(ParseConfig(config_path));
+    wds_shards_prefixsums_.push_back(total_size_);
+    total_size_ += wds_shards_metadata_.back().size();
+  }
+  wds_shards_prefixsums_.push_back(total_size_);  // for the last shard when reaches the end
+
+  // initializing all the readers
+  for (auto& uri : uris_) {
+    wds_shards_.emplace_back(FileStream::Open(uri, read_ahead_, !dont_use_mmap_));
+  }
+
+  // seeking the first wds shard to use
+  const size_t first_index_ = start_index(shard_id_, num_shards_, total_size_);
+
+  while (wds_shards_prefixsums_[first_wds_shard_index_] +
+             wds_shards_metadata_[first_wds_shard_index_].size() <=
+         first_sample_index_) {
+    first_wds_shard_index_++;
+  }
+  first_sample_index_ = first_sample_index_ - wds_shards_prefixsums_[first_wds_shard_index_];
+
+  if (stick_to_shard_) {
+    current_wds_shard_index_ = first_wds_shard_index_;
+    current_sample_index_ = first_sample_index_;
+  }
+
+  // initializing the first reader
+  if (stick_to_shard_) {
+    wds_shards_[first_wds_shard_index_].SeekArchive(
+        wds_shards_metadata_[first_wds_shard_index_][first_sample_index_].start_offset);
+  }
+}
+
+void WebdatasetLoader::Reset(bool wrap_to_shard) {
+  current_wds_shard_index_ = wrap_to_shard ? first_wds_shard_index_ : 0;
+  current_sample_index_ = wrap_to_shard ? first_sample_index_ : 0;
+  for (detail::TarArchive& wds_shard : wds_shards_) {
+    wds_shard.SeekArchive(0);
+  }
```

Comment:
L184 handles that. This is not needed here I guess.

Reply:
removed in the current changes, will push it tomorrow